crates/xilem_html/web_examples/todomvc/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+                    .on_click(|state: &mut AppState, _| {
+                        state.filter = Filter::All;
+                        MessageResult::RequestRebuild
+                    }),
+            ),
+            text(" "),
+            el::li(
+                el::a(text("Active"))
+                    .attr("href", "#/active")
+                    .attr("class", filter_class(Filter::Active))
+                    .on_click(|state: &mut AppState, _| {
+                        state.filter = Filter::Active;
+                        MessageResult::RequestRebuild
+                    }),
+            ),
+            text(" "),
+            el::li(
+                el::a(text("Completed"))
+                    .attr("href", "#/completed")
+                    .attr("class", filter_class(Filter::Completed))
+                    .on_click(|state: &mut AppState, _| {
+                        state.filter = Filter::Completed;
+                        MessageResult::RequestRebuild
+                    }),
+            ),
+        ))
+        .attr("class", "filters"),
+        clear_button,
+    ))
+    .attr("class", "footer")
+}
+
+fn main_view(state: &mut AppState) -> impl View<AppState> + ViewMarker {
+    let editing_id = state.editing_id;
+    let todos: Vec<_> = state
+        .visible_todos()
+        .map(|(idx, todo)| {
+            Adapt::new(
+                move |data: &mut AppState, thunk| {
+                    if let MessageResult::Action(action) = thunk.call(&mut data.todos[idx]) {
+                        match action {
+                            TodoAction::SetEditing(id) => data.start_editing(id),
+                            TodoAction::CancelEditing => data.editing_id = None,
+                            TodoAction::Destroy(id) => data.todos.retain(|todo| todo.id != id),
+                        }
+                    }
+                    MessageResult::Nop
+                },
+                todo_item(todo, editing_id == Some(todo.id)),
+            )
+        })
+        .collect();
+    el::section((
+        el::input(())
+            .attr("id", "toggle-all")
+            .attr("class", "toggle-all")
+            .attr("type", "checkbox")
+            .attr("checked", "true"),
+        el::label(()).attr("for", "toggle-all"),
+        el::ul(todos).attr("class", "todo-list"),
+    ))
+    .attr("class", "main")
+}
+
+fn app_logic(state: &mut AppState) -> impl View<AppState> {
+    log::debug!("render: {state:?}");
+    let main = (!state.todos.is_empty()).then(|| main_view(state));
+    let footer = (!state.todos.is_empty()).then(|| footer_view(state));
+    el::div((
+        el::header((
+            el::h1(text("TODOs")),
+            el::input(())
+                .attr("class", "new-todo")
+                .attr("placeholder", "What needs to be done?")
+                .attr("value", state.new_todo.clone())
+                .attr("autofocus", "true")
+                .on_keydown(|state: &mut AppState, evt| {
```

Comment:
My feeling is that these callbacks should be returning an `A` rather than a raw `MessageResult`, as the latter potentially exposes too many internal details. But this is open for discussion.

I don't think we need an explicit RequestRebuild (the intent of that is to propagate changes when an async future completes, and in particular blasts through memoize or other lazy mechanisms). The re-render of the view tree should happen naturally after a message is delivered.

Reply:
I've changed the callbacks to use one of `()`, `A` or `Option<A>`. Unfortunately this does require a trait bound on `A` (I made a marker trait called Action). An alternative would be to force the callback to return an `Option<A>`, in which case that restriction could be relaxed. The reason for needing the bound is to ensure the implementatinos don't overlap, in particular `Action` is not implemented for `()`.

Reply:
I haven't changed it on the `Adapt` view because I didn't work out how to do it in a backend-agnostic way.